src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNITcpHandle.cs

```diff
@@ -100,86 +105,135 @@ public override uint Status
+                CancellationToken.None,
+                TaskCreationOptions.DenyChildAttach,
+                TaskScheduler.Default
+            ).Unwrap();
+        }
 
+        private static async Task<SNIError> BeginOpenConnectionAsync(object state)
+        {
+            var parameters = (Tuple<SNITCPHandle, int, long, bool>)state;
+            SNITCPHandle instance = parameters.Item1;
+            int port = parameters.Item2;
+            long timerExpire = parameters.Item3;
+            bool parallel = parameters.Item4;
+
+            SNIError retval = null;
             try
             {
-                TimeSpan ts = default(TimeSpan);
+                TimeSpan timeout = default;
 
                 // In case the Timeout is Infinite, we will receive the max value of Int64 as the tick count
                 // The infinite Timeout is a function of ConnectionString Timeout=0
                 bool isInfiniteTimeOut = long.MaxValue == timerExpire;
                 if (!isInfiniteTimeOut)
                 {
-                    ts = DateTime.FromFileTime(timerExpire) - DateTime.Now;
-                    ts = ts.Ticks < 0 ? TimeSpan.FromTicks(0) : ts;
+                    timeout = DateTime.FromFileTime(timerExpire) - DateTime.Now;
+                    timeout = timeout.Ticks < 0 ? TimeSpan.FromTicks(0) : timeout;
                 }
 
                 Task<Socket> connectTask;
                 if (parallel)
                 {
-                    Task<IPAddress[]> serverAddrTask = Dns.GetHostAddressesAsync(serverName);
-                    serverAddrTask.Wait(ts);
-                    IPAddress[] serverAddresses = serverAddrTask.Result;
+                    IPAddress[] serverAddresses = await Dns.GetHostAddressesAsync(instance._targetServer);
 
                     if (serverAddresses.Length > MaxParallelIpAddresses)
                     {
                         // Fail if above 64 to match legacy behavior
-                        ReportTcpSNIError(0, SNICommon.MultiSubnetFailoverWithMoreThan64IPs, string.Empty);
-                        return;
+                        return CreateTcpSNIError(0, SNICommon.MultiSubnetFailoverWithMoreThan64IPs, string.Empty);
                     }
 
                     connectTask = ParallelConnectAsync(serverAddresses, port);
-
-                    if (!(isInfiniteTimeOut ? connectTask.Wait(-1) : connectTask.Wait(ts)))
+                    if (!connectTask.IsCompleted)
                     {
-                        ReportTcpSNIError(0, SNICommon.ConnOpenFailedError, string.Empty);
-                        return;
+                        CancellationTokenSource cancellationTokenSource = new CancellationTokenSource();
+                        if (!isInfiniteTimeOut)
+                        {
+                            cancellationTokenSource.CancelAfter(timeout);
+                        }
+                        connectTask.Wait(cancellationTokenSource.Token);
```

Comment:
Sync Wait in async method? Can't you just `await` it?

Also you could pickup one of the Timeout patterns from the TaskTimeoutExtensions in tests https://github.com/dotnet/corefx/blob/407c09d5840384fbd24bbd5b8869ef13be71de63/src/Common/tests/System/Threading/Tasks/TaskTimeoutExtensions.cs#L28-L43

/cc @stephentoub

Reply:
I can't test any of the parallel branch and the code in the ParallelAsyncHelper is a bit beyond me. If there's a way to safely await it and preserve behaviour without risk i'm open to suggestions.
Similarly if there's a good way to avoid the async wait in ResolveConnectionStatus it would be good but I don't see one without an async lock.